optaplanner-quarkus-integration/optaplanner-quarkus/deployment/src/main/java/org/optaplanner/quarkus/deployment/OptaPlannerProcessor.java

```diff
@@ -211,31 +214,77 @@ private void applySolverProperties(RecorderContext recorderContext,
     }
 
     private void applyScoreDirectorFactoryProperties(IndexView indexView, SolverConfig solverConfig) {
+        Optional<String> constraintsDrlFromProperty = constraintsDrl();
+        Optional<String> defaultConstraintsDrl = defaultConstraintsDrl();
+        Optional<String> effectiveConstraintsDrl = constraintsDrlFromProperty.map(Optional::of).orElse(defaultConstraintsDrl);
         if (solverConfig.getScoreDirectorFactoryConfig() == null) {
-            ScoreDirectorFactoryConfig scoreDirectorFactoryConfig = new ScoreDirectorFactoryConfig();
-            scoreDirectorFactoryConfig.setEasyScoreCalculatorClass(
-                    findImplementingClass(DotNames.EASY_SCORE_CALCULATOR, indexView));
-            scoreDirectorFactoryConfig.setConstraintProviderClass(
-                    findImplementingClass(DotNames.CONSTRAINT_PROVIDER, indexView));
-            scoreDirectorFactoryConfig.setIncrementalScoreCalculatorClass(
-                    findImplementingClass(DotNames.INCREMENTAL_SCORE_CALCULATOR, indexView));
+            ScoreDirectorFactoryConfig scoreDirectorFactoryConfig =
+                    defaultScoreDirectoryFactoryConfig(indexView, effectiveConstraintsDrl);
+            solverConfig.setScoreDirectorFactoryConfig(scoreDirectorFactoryConfig);
+        } else {
+            ScoreDirectorFactoryConfig scoreDirectorFactoryConfig = solverConfig.getScoreDirectorFactoryConfig();
+            if (constraintsDrlFromProperty.isPresent()) {
+                scoreDirectorFactoryConfig.setScoreDrlList(Collections.singletonList(constraintsDrlFromProperty.get()));
+            } else {
+                if (scoreDirectorFactoryConfig.getScoreDrlList() == null) {
+                    defaultConstraintsDrl.ifPresent((resolvedConstraintsDrl) -> scoreDirectorFactoryConfig
+                            .setScoreDrlList(Collections.singletonList(resolvedConstraintsDrl)));
+                }
+            }
+        }
+
+        if (solverConfig.getScoreDirectorFactoryConfig().getScoreDrlList() != null) {
             ClassLoader classLoader = Thread.currentThread().getContextClassLoader();
-            if (classLoader.getResource(SolverBuildTimeConfig.DEFAULT_SCORE_DRL_URL) != null) {
-                scoreDirectorFactoryConfig.setScoreDrlList(Collections.singletonList(
-                        SolverBuildTimeConfig.DEFAULT_SCORE_DRL_URL));
+            try {
+                Class.forName("org.drools.dynamic.DynamicComponentsSupplier", false, classLoader);
+            } catch (ClassNotFoundException e) {
+                throw new IllegalStateException(
+                        "Constraints DRL has been detected, but drools-core-dynamic is not on classpath."
+                                + " Maybe try adding it. Please note that constraints DRL does not work in the native mode.");
```

Comment:
I tend to avoid "it" references in the "error maybe fix" to the previous line, so users can "just fix it and move on" faster. Especially if the "error fact" is very long and complex to parse (hard to see in a console window, especially if it doesn't wrap and because the error message only starts at column 70 or so)
```suggestion
                                + "\nMaybe add the dependency org.kie.kogito:drools-core-dynamic and exclude the dependency org.kie.kogito:drools-core-static.");
```

Reply:
For the part about not working in native mode: that doesn't belong here, but in the Quarkus integration chapter docs, especially on the constraints-drl property.
Also, if the do try to run it in native mode with a scoreDRL, the error message should be proper - but that's a different check than this one - can we somehow detect that?

Reply:
Also, what's the error message if I add dynamic to the classpath but forget to exclude static?

Reply:
Agreed: let's provide fail fast checks for both.

"Maybe" suggestion for native: either use ConstraintProvider or don't use native.